Reuse commit messages that failed previous style check

all3fox opened the pull request:
This PR is a work-in-progress about improving feedback flow as discussed in #85 

Currently, the base case works: if style check fails, the message is saved in `COMMIT_EDITMSG_OLD` and suggested to the user during the next commit attempt.

However, this needs some thought/testing before merging. Here is an example scenario that causes problems now:

1. The user does a `git commit --amend`
2. The message they enter fails validation, the commit does not happen.
3. The user decides they do not want an amend after all, they revert all the work and forget about it.
4. After a while, the user makes a completely different contribution.
5. When they do a `git commit`, they get an old commit message that failed long ago
6. Confusion?

Or maybe I am just overthinking it, let me know.

jzaefferer commented:
How about checking the time COMMIT_EDITMSG_OLD was last modified, and only using if it its less than 5 minutes? Or 10? Anything longer and the user is likely at the point where they forgot about it.

all3fox commented:
Sure, up for review now. The message is not reused if it is > 300 seconds old

jzaefferer commented on install.js:
```diff
@@ -35,26 +35,47 @@ var hooks = path.join(git, 'hooks')
+
+var i, dstHook, srcHook
+
+for (i = 0; i < dstHooks.length; ++i) {
```
Could use `dstHooks.forEach`

jzaefferer commented on install.js:
```diff
@@ -35,26 +35,47 @@ var hooks = path.join(git, 'hooks')
-      chalk.red(
-        'Failed to write commit-msg hook. ' +
-        'Make sure you have the necessary permissions.'
+for (i = 0; i < dstHooks.length; ++i) {
```
Same here, though you'd still need the second array access.

jzaefferer commented on prepare-commit-msg-hook.js:
```diff
@@ -0,0 +1,26 @@
+
+  var mtime = new Date(fs.statSync(oldMessagePath).mtime)
+
+  // Date.now() - mtime.getTime() is milliseconds, convert to minutes
```
"convert to minutes"? Looks like you're converting to seconds now.

jzaefferer commented on prepare-commit-msg-hook.js:
```diff
@@ -0,0 +1,26 @@
+  // Date.now() - mtime.getTime() is milliseconds, convert to minutes
+  if ((Date.now() - mtime.getTime()) / 1000 < oldMessageSeconds) {
+    fs.writeFileSync(
+      process.argv[2], fs.readFileSync(oldMessagePath).toString()
```
tiny optimization, maybe: is the toString() needed here? writeFileSync should be able to write a buffer

jzaefferer commented on prepare-commit-msg-hook.js:
```diff
@@ -0,0 +1,26 @@
+  // There is an old message that was previously rejected by us
+  // Suggest it to the user so they do not have to start from scratch
+
+  var mtime = new Date(fs.statSync(oldMessagePath).mtime)
```
Do we need `existsSync` if we're doing `statSync` anyway?

all3fox commented:
Thanks for the review! I have addressed all of those, you should probably look again at the `prepare-commit-msg-hook.js`, where I changed `if (...)` for a `try {} catch ()`.

And uninstall.js got new functionality (remove old message file if you uninstall commitplease)

jzaefferer commented on install.js:
```diff
@@ -35,26 +35,45 @@ var hooks = path.join(git, 'hooks')
-        'Failed to write commit-msg hook. ' +
-        'Make sure you have the necessary permissions.'
+var i = 0
+dstHooks.forEach(function (dstHook) {
```
The 2nd argument in a forEach callback is the index. You can use that to replace your custom counter.

all3fox commented:
Thanks, I have fixed that, will now squash stuff together and merge to master, bumping the version to `2.7.8`

all3fox commented:
Ok, I've merged and pushed the new version/its tag to git. However, a small hiccup happened when I tried to `npm publish`:

```
alex@tardis ~/Code/commitplease (git)-[master] % npm publish            
npm ERR! registry error parsing json
+ commitplease@2.7.8
alex@tardis ~/Code/commitplease (git)-[master] % git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working tree clean
alex@tardis ~/Code/commitplease (git)-[master] % npm publish
npm ERR! publish Failed PUT 403
npm ERR! Linux 4.9.8-1-ARCH
npm ERR! argv "/home/alex/.nvm/versions/node/v6.5.0/bin/node" "/home/alex/.nvm/versions/node/v6.5.0/bin/npm" "publish"
npm ERR! node v6.5.0
npm ERR! npm  v3.10.3
npm ERR! code E403

npm ERR! "You cannot publish over the previously published version 2.7.8." : commitplease
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /home/alex/Code/commitplease/npm-debug.log
```

Basically, my first `npm publish` failed and my second `npm publish` told me that `commitplease` is already 2.7.8

`npm install commitplease` installs the `2.7.8` version as it should, so it all seems to work.

jzaefferer commented:
Since this is a new feature, a minor version bump would've been fine. I guess its too late for that.

Anyway, nice work! Great to have this resolved. I'll close the parent issue.
